clang-plugin/MozsearchIndexer.cpp

```diff
@@ -1138,6 +1138,33 @@ class IndexConsumer : public ASTConsumer,
     return CurDeclContext && CurDeclContext->VisitImplicit;
   }
 
+  // We don't want to traverse all specializations everytime we find a forward
+  // declaration, so only traverse specializations related to an actual
+  // definition.
+  //
+  // For nested templates, we have to check if the original MemberTemplate is is
+  // a definition, to handle cases such as:
+  // ```
+  // template <typename T>
+  // struct Maybe {
+  //   template <typename... Args>
+  //   T *emplace(Args&&... args) {
+  //     defined here
+  //   }
+  // }
+  // ```
```

Comment:
Should this example code include the forward declaration from https://bugzilla.mozilla.org/show_bug.cgi?id=1933325#c1 potentially with a comment on the forward like: "This is a canonical declaration, but not a definition."?

(Also this is a really great comment, thanks!)

Reply:
While getting the std::vector::emplace_back working, I think I got a better understanding of what is going on here, and settled on isThisDeclarationADefinition for class templates and keeping isCanonicalDecl for function templates. I've updated and completed the comment in that regard.
I'll admit I'm still not 100% confident in my understanding of this though :sweat_smile:

Reply:
Ha, yeah, when it comes to templates I very much appreciate that we have a test corpus with precisely targeted coverage because everything about templates tends to test the limits of my mental stack and my reviews are very much of a "this code does not contradict my mental model of what's going on" where my mental model frequently regresses to "templates are spicy macros".